### PR TITLE
[dark canary] added the ability to pass in a upstreamCallback function

### PR DESCRIPTION
**Pull request by JohnnyJohnAndTheFunkyBunch**
I added the ability to pass in a callback function in DarkCanaryChannelHandler so that clients can receive the responses from the host or its mirror. In doing so I added a few HashMap into the logic, which I made sure it wouldn't cause memory leak by also having a cleanup for host requests.

Note: In JMX.scala, I uncommented out the code. because I believe it affected cloud-ncs, and it shouldn't be removed.

### CODE REVIEW

File: network/src/main/scala/com/linkedin/norbert/network/client/NetworkClient.scala [hunk 1]
```diff
@@ -62,6 +63,7 @@ class NetworkClientConfig {
 
   var avoidByteStringCopy = NetworkDefaults.AVOID_BYTESTRING_COPY
   var darkCanaryServiceName: Option[String] = None
+  var darkCanaryUpstreamCallback: Option[(Boolean,UUID,Object)=>Unit] = None
```
Comment (sungjuc) [inline network/src/main/scala/com/linkedin/norbert/network/client/NetworkClient.scala]: Please add spaces.

File: network/src/main/scala/com/linkedin/norbert/network/netty/DarkCanaryChannelHandler.scala [hunk 2]
```diff
@@ -153,7 +158,13 @@ class DarkCanaryChannelHandler extends Logging {
                     None,
                     0)
 
-                  requestMap.put(newRequest.id, newRequest)
+                  if (!upstreamCallback.isEmpty) {
```
Comment (sungjuc) [inline network/src/main/scala/com/linkedin/norbert/network/netty/DarkCanaryChannelHandler.scala]: upstreamCallback is an Option. Please compare with None or Some

File: network/src/main/scala/com/linkedin/norbert/network/netty/DarkCanaryChannelHandler.scala [hunk 3]
```diff
@@ -174,15 +185,25 @@ class DarkCanaryChannelHandler extends Logging {
...
               case request: Request[Any,Any] =>  {
-                requestMap.remove(requestId)
+                mirrorRequestMap.remove(requestId)
+                // Begin upstreamCallback
+                if (!upstreamCallback.isEmpty) {
+                  mirrorToHost.remove(requestId) match {
+                    case hostRequestId => {
+                      upstreamCallback.get(true, hostRequestId, message)
+                    }
+                    case null => log.error("Could not find hostRequestId for darkRequestId: %s".format(requestId.toString))
```
Comment (sungjuc) [inline network/src/main/scala/com/linkedin/norbert/network/netty/DarkCanaryChannelHandler.scala]: darkReuqestId => darkCanaryRequestId

File: cluster/src/main/scala/com/linkedin/norbert/jmx/JMX.scala [hunk 4]
```diff
@@ -51,7 +51,7 @@ object JMX extends Logging {
 
   def unregister(mbean: ObjectInstance) = try {
     mbeanServer.unregisterMBean(mbean.getObjectName)
-    //synchronized { map.remove(mbean.getObjectName.getCanonicalName) } We treat the map as a mapping
+    synchronized { map.remove(mbean.getObjectName.getCanonicalName) } //We treat the map as a mapping
```
Comment (shyamshankar) [inline cluster/src/main/scala/com/linkedin/norbert/jmx/JMX.scala]: Why uncomment this?
This doesn't seem to be related to our change.

Comment (JohnnyJohnAndTheFunkyBunch) [inline cluster/src/main/scala/com/linkedin/norbert/jmx/JMX.scala]: cloud-ncs dails if i dont uncomment this. I pulled the latest norbert, and cloud ncs doesn't pass its tests.

Comment (sungjuc) [inline cluster/src/main/scala/com/linkedin/norbert/jmx/JMX.scala]: I think the commenting is required. Previous change mistakenly commented out it.
